### DISCUSSION TITLE
Missing vulnerabilities in check

### DISCUSSION THREAD

**Discussion by savujevi** [DependencyTrack/dependency-track#1140] [Q&A]
All questions should be directed to the Dependency-Track Slack channel or to GitHub discussions
Questions submitted through GitHub issues will be closed.

Slack channel
  - https://dependencytrack.org/slack

Slack invite
  - https://dependencytrack.org/slack/invite
  
GitHub discussions
  - https://github.com/DependencyTrack/dependency-track/discussions

Groups.io (mailing list)
  - https://dependencytrack.org/discussion

Providing as much detail as possible, what is your question?

Hello developers,

I am using the latest version of Dependenvytrack (api and frontend) and I am checking a maven project and its bom.
When I compare the result with the result of the tool dependency-ckeck from OWASP I am missing some vulnerabilities.

Maybe you can tell me where the difference is in the dependency-ckeck and the dependencytrack.

Here is the dependency tree and the missing vulnerabilities in the result of the dependencytrack. I would share the bom.json but can't attache it.

commons-io-2.5.jar	                                 NVD CVE-2021-29425  Medium  CWE-22
spring-core-5.2.9.RELEASE.jar	                 NVD CVE-2021-22118	High    CWE-269
spring-security-core-5.3.4.RELEASE.jar	NVD CVE-2021-22118	High    CWE-269
spring-security-core-5.3.4.RELEASE.jar	NVD CVE-2021-22119	High    CWE-863
spring-security-web-5.3.4.RELEASE.jar	NVD CVE-2021-22119	High    CWE-863

[INFO] --- maven-dependency-plugin:3.1.2:tree (default-cli) @ partnerauskunft ---
[INFO] de.ergodirekt.partner.microservice:partnerauskunft:jar:2.0.3-SNAPSHOT
[INFO] +- de.ergodirekt.partner:kerberos-delegation-lib:jar:1.0.3:compile
[INFO] +- de.ergodirekt.partner:partner-compliance-lib:jar:1.0.7:compile
[INFO] +- org.springframework.boot:spring-boot-starter-hateoas:jar:2.3.4.RELEASE:compile
[INFO] |  \- org.springframework.hateoas:spring-hateoas:jar:1.1.2.RELEASE:compile
[INFO] |     +- org.springframework:spring-beans:jar:5.2.9.RELEASE:compile
[INFO] |     \- org.springframework:spring-context:jar:5.2.9.RELEASE:compile
[INFO] +- org.springframework.boot:spring-boot-starter-web:jar:2.3.4.RELEASE:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter:jar:2.3.4.RELEASE:compile
[INFO] |  |  +- org.springframework.boot:spring-boot:jar:2.3.4.RELEASE:compile
[INFO] |  |  +- org.springframework.boot:spring-boot-autoconfigure:jar:2.3.4.RELEASE:compile
[INFO] |  |  +- org.springframework.boot:spring-boot-starter-logging:jar:2.3.4.RELEASE:compile
[INFO] |  |  |  +- ch.qos.logback:logback-classic:jar:1.2.3:compile
[INFO] |  |  |  +- org.apache.logging.log4j:log4j-to-slf4j:jar:2.13.3:compile
[INFO] |  |  |  |  \- org.apache.logging.log4j:log4j-api:jar:2.13.3:compile
[INFO] |  |  |  \- org.slf4j:jul-to-slf4j:jar:1.7.30:compile
[INFO] |  |  +- jakarta.annotation:jakarta.annotation-api:jar:1.3.5:compile
[INFO] |  |  \- org.yaml:snakeyaml:jar:1.26:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter-json:jar:2.3.4.RELEASE:compile
[INFO] |  |  +- com.fasterxml.jackson.datatype:jackson-datatype-jdk8:jar:2.11.2:compile
[INFO] |  |  \- com.fasterxml.jackson.module:jackson-module-parameter-names:jar:2.11.2:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter-tomcat:jar:2.3.4.RELEASE:compile
[INFO] |  |  +- org.apache.tomcat.embed:tomcat-embed-core:jar:9.0.38:compile
[INFO] |  |  +- org.glassfish:jakarta.el:jar:3.0.3:compile
[INFO] |  |  \- org.apache.tomcat.embed:tomcat-embed-websocket:jar:9.0.38:compile
[INFO] |  +- org.springframework:spring-web:jar:5.2.9.RELEASE:compile
[INFO] |  \- org.springframework:spring-webmvc:jar:5.2.9.RELEASE:compile
[INFO] |     \- org.springframework:spring-expression:jar:5.2.9.RELEASE:compile
[INFO] +- org.springframework.boot:spring-boot-starter-test:jar:2.3.4.RELEASE:test
[INFO] |  +- org.springframework.boot:spring-boot-test:jar:2.3.4.RELEASE:test
[INFO] |  +- org.springframework.boot:spring-boot-test-autoconfigure:jar:2.3.4.RELEASE:test
[INFO] |  +- com.jayway.jsonpath:json-path:jar:2.4.0:compile
[INFO] |  |  \- net.minidev:json-smart:jar:2.3:compile
[INFO] |  |     \- net.minidev:accessors-smart:jar:1.2:compile
[INFO] |  |        \- org.ow2.asm:asm:jar:5.0.4:compile
[INFO] |  +- jakarta.xml.bind:jakarta.xml.bind-api:jar:2.3.3:compile
[INFO] |  |  \- jakarta.activation:jakarta.activation-api:jar:1.2.2:compile
[INFO] |  +- org.assertj:assertj-core:jar:3.16.1:test
[INFO] |  +- org.hamcrest:hamcrest:jar:2.2:test
[INFO] |  +- org.junit.jupiter:junit-jupiter:jar:5.7.0:test
[INFO] |  |  +- org.junit.jupiter:junit-jupiter-api:jar:5.6.2:test
[INFO] |  |  |  +- org.opentest4j:opentest4j:jar:1.2.0:test
[INFO] |  |  |  \- org.junit.platform:junit-platform-commons:jar:1.6.2:test
[INFO] |  |  +- org.junit.jupiter:junit-jupiter-params:jar:5.6.2:test
[INFO] |  |  \- org.junit.jupiter:junit-jupiter-engine:jar:5.6.2:test
[INFO] |  +- org.junit.vintage:junit-vintage-engine:jar:5.7.0:test
[INFO] |  |  +- org.apiguardian:apiguardian-api:jar:1.1.0:test
[INFO] |  |  \- org.junit.platform:junit-platform-engine:jar:1.7.0:test
[INFO] |  +- org.mockito:mockito-core:jar:3.3.3:test
[INFO] |  |  +- net.bytebuddy:byte-buddy:jar:1.10.14:compile
[INFO] |  |  +- net.bytebuddy:byte-buddy-agent:jar:1.10.14:test
[INFO] |  |  \- org.objenesis:objenesis:jar:2.6:test
[INFO] |  +- org.mockito:mockito-junit-jupiter:jar:3.3.3:test
[INFO] |  +- org.springframework:spring-core:jar:5.2.9.RELEASE:compile
[INFO] |  |  \- org.springframework:spring-jcl:jar:5.2.9.RELEASE:compile
[INFO] |  +- org.springframework:spring-test:jar:5.2.9.RELEASE:test
[INFO] |  \- org.xmlunit:xmlunit-core:jar:2.7.0:test
[INFO] +- org.springframework.boot:spring-boot-starter-data-jpa:jar:2.3.4.RELEASE:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter-aop:jar:2.3.4.RELEASE:compile
[INFO] |  |  \- org.aspectj:aspectjweaver:jar:1.9.6:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter-jdbc:jar:2.3.4.RELEASE:compile
[INFO] |  |  +- com.zaxxer:HikariCP:jar:3.4.5:compile
[INFO] |  |  \- org.springframework:spring-jdbc:jar:5.2.9.RELEASE:compile
[INFO] |  +- jakarta.transaction:jakarta.transaction-api:jar:1.3.3:compile
[INFO] |  +- jakarta.persistence:jakarta.persistence-api:jar:2.2.3:compile
[INFO] |  +- org.hibernate:hibernate-core:jar:5.4.21.Final:compile
[INFO] |  |  +- org.jboss.logging:jboss-logging:jar:3.4.1.Final:compile
[INFO] |  |  +- org.javassist:javassist:jar:3.24.0-GA:compile
[INFO] |  |  +- antlr:antlr:jar:2.7.7:compile
[INFO] |  |  +- org.jboss:jandex:jar:2.1.3.Final:compile
[INFO] |  |  +- org.dom4j:dom4j:jar:2.1.3:compile
[INFO] |  |  +- org.hibernate.common:hibernate-commons-annotations:jar:5.1.0.Final:compile
[INFO] |  |  \- org.glassfish.jaxb:jaxb-runtime:jar:2.3.3:compile
[INFO] |  |     +- org.glassfish.jaxb:txw2:jar:2.3.3:compile
[INFO] |  |     +- com.sun.istack:istack-commons-runtime:jar:3.0.11:compile
[INFO] |  |     \- com.sun.activation:jakarta.activation:jar:1.2.2:runtime
[INFO] |  +- org.springframework.data:spring-data-jpa:jar:2.3.4.RELEASE:compile
[INFO] |  |  +- org.springframework.data:spring-data-commons:jar:2.3.4.RELEASE:compile
[INFO] |  |  +- org.springframework:spring-orm:jar:5.2.9.RELEASE:compile
[INFO] |  |  \- org.springframework:spring-tx:jar:5.2.9.RELEASE:compile
[INFO] |  \- org.springframework:spring-aspects:jar:5.2.9.RELEASE:compile
[INFO] +- com.ibm.db2:db2jcc:jar:4.14.88:compile
[INFO] +- com.ibm.db2:db2jcc_license_cisuz:jar:4.14.88:compile
[INFO] +- org.skyscreamer:jsonassert:jar:1.5.0:compile
[INFO] |  \- com.vaadin.external.google:android-json:jar:0.0.20131108.vaadin1:compile
[INFO] +- org.apache.commons:commons-lang3:jar:3.11:compile
[INFO] +- commons-io:commons-io:jar:2.5:compile
[INFO] +- org.unitils:unitils-core:jar:3.4.3:compile
[INFO] |  +- junit:junit:jar:4.13:compile
[INFO] |  |  \- org.hamcrest:hamcrest-core:jar:2.2:test
[INFO] |  +- commons-logging:commons-logging:jar:1.1:compile
[INFO] |  +- commons-lang:commons-lang:jar:2.3:compile
[INFO] |  +- commons-collections:commons-collections:jar:3.2:compile
[INFO] |  \- ognl:ognl:jar:2.6.9:compile
[INFO] +- org.springframework.boot:spring-boot-starter-actuator:jar:2.3.4.RELEASE:compile
[INFO] |  +- org.springframework.boot:spring-boot-actuator-autoconfigure:jar:2.3.4.RELEASE:compile
[INFO] |  |  \- org.springframework.boot:spring-boot-actuator:jar:2.3.4.RELEASE:compile
[INFO] |  \- io.micrometer:micrometer-core:jar:1.5.5:compile
[INFO] |     +- org.hdrhistogram:HdrHistogram:jar:2.1.12:compile
[INFO] |     \- org.latencyutils:LatencyUtils:jar:2.0.3:runtime
[INFO] +- org.springframework.boot:spring-boot-starter-security:jar:2.3.4.RELEASE:compile
[INFO] |  +- org.springframework:spring-aop:jar:5.2.9.RELEASE:compile
[INFO] |  +- org.springframework.security:spring-security-config:jar:5.3.4.RELEASE:compile
[INFO] |  \- org.springframework.security:spring-security-web:jar:5.3.4.RELEASE:compile
[INFO] +- org.springframework.boot:spring-boot-starter-data-ldap:jar:2.3.4.RELEASE:compile
[INFO] |  \- org.springframework.data:spring-data-ldap:jar:2.3.4.RELEASE:compile
[INFO] |     \- org.springframework.ldap:spring-ldap-core:jar:2.3.3.RELEASE:compile
[INFO] +- org.springframework.security:spring-security-test:jar:5.3.4.RELEASE:test
[INFO] |  \- org.springframework.security:spring-security-core:jar:5.3.4.RELEASE:compile
[INFO] +- org.springframework.security.kerberos:spring-security-kerberos-web:jar:1.0.1.RELEASE:compile
[INFO] |  \- org.springframework.security.kerberos:spring-security-kerberos-core:jar:1.0.1.RELEASE:compile
[INFO] +- io.springfox:springfox-boot-starter:jar:3.0.0:compile
[INFO] |  +- io.springfox:springfox-oas:jar:3.0.0:compile
[INFO] |  |  +- io.swagger.core.v3:swagger-annotations:jar:2.1.2:compile
[INFO] |  |  +- io.swagger.core.v3:swagger-models:jar:2.1.2:compile
[INFO] |  |  +- io.springfox:springfox-spi:jar:3.0.0:compile
[INFO] |  |  +- io.springfox:springfox-schema:jar:3.0.0:compile
[INFO] |  |  +- io.springfox:springfox-core:jar:3.0.0:compile
[INFO] |  |  +- io.springfox:springfox-spring-web:jar:3.0.0:compile
[INFO] |  |  |  \- io.github.classgraph:classgraph:jar:4.8.83:compile
[INFO] |  |  +- io.springfox:springfox-spring-webmvc:jar:3.0.0:compile
[INFO] |  |  +- io.springfox:springfox-spring-webflux:jar:3.0.0:compile
[INFO] |  |  +- io.springfox:springfox-swagger-common:jar:3.0.0:compile
[INFO] |  |  \- org.mapstruct:mapstruct:jar:1.3.1.Final:runtime
[INFO] |  +- io.springfox:springfox-data-rest:jar:3.0.0:compile
[INFO] |  +- io.springfox:springfox-bean-validators:jar:3.0.0:compile
[INFO] |  +- io.springfox:springfox-swagger2:jar:3.0.0:compile
[INFO] |  |  +- io.swagger:swagger-annotations:jar:1.5.20:compile
[INFO] |  |  \- io.swagger:swagger-models:jar:1.5.20:compile
[INFO] |  +- io.springfox:springfox-swagger-ui:jar:3.0.0:compile
[INFO] |  +- com.fasterxml:classmate:jar:1.5.1:compile
[INFO] |  +- org.slf4j:slf4j-api:jar:1.7.30:compile
[INFO] |  +- org.springframework.plugin:spring-plugin-core:jar:2.0.0.RELEASE:compile
[INFO] |  \- org.springframework.plugin:spring-plugin-metadata:jar:2.0.0.RELEASE:compile
[INFO] +- com.fasterxml.jackson.datatype:jackson-datatype-hibernate5:jar:2.11.2:compile
[INFO] |  +- javax.transaction:javax.transaction-api:jar:1.3:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.11.2:compile
[INFO] |  \- com.fasterxml.jackson.core:jackson-databind:jar:2.11.2:compile
[INFO] +- com.fasterxml.jackson.module:jackson-module-afterburner:jar:2.11.2:compile
[INFO] +- com.fasterxml.jackson.dataformat:jackson-dataformat-xml:jar:2.11.2:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.11.2:compile
[INFO] |  +- com.fasterxml.jackson.module:jackson-module-jaxb-annotations:jar:2.11.2:compile
[INFO] |  +- org.codehaus.woodstox:stax2-api:jar:4.2.1:compile
[INFO] |  \- com.fasterxml.woodstox:woodstox-core:jar:6.2.1:compile
[INFO] +- com.fasterxml.jackson.datatype:jackson-datatype-jsr310:jar:2.11.2:compile
[INFO] +- com.fsp.org:org-javaapi:jar:3.1.7:compile
[INFO] +- org.apache.httpcomponents:httpclient:jar:4.5.12:test
[INFO] |  +- org.apache.httpcomponents:httpcore:jar:4.4.13:test
[INFO] |  \- commons-codec:commons-codec:jar:1.14:test
[INFO] +- org.zalando:problem-spring-web:jar:0.20.1:compile
[INFO] |  +- com.google.code.findbugs:jsr305:jar:3.0.2:compile
[INFO] |  +- org.zalando:problem:jar:0.20.0:compile
[INFO] |  |  \- com.google.gag:gag:jar:1.0.1:compile
[INFO] |  +- org.zalando:jackson-datatype-problem:jar:0.20.0:compile
[INFO] |  +- org.zalando:faux-pas:jar:0.4.0:compile
[INFO] |  \- javax.validation:validation-api:jar:2.0.1.Final:compile
[INFO] +- net.logstash.logback:logstash-logback-encoder:jar:5.1:compile
[INFO] |  \- ch.qos.logback:logback-core:jar:1.2.3:compile
[INFO] \- com.google.guava:guava:jar:28.2-jre:compile
[INFO]    +- com.google.guava:failureaccess:jar:1.0.1:compile
[INFO]    +- com.google.guava:listenablefuture:jar:9999.0-empty-to-avoid-conflict-with-guava:compile
[INFO]    +- org.checkerframework:checker-qual:jar:2.10.0:compile
[INFO]    +- com.google.errorprone:error_prone_annotations:jar:2.3.4:compile
[INFO]    \- com.google.j2objc:j2objc-annotations:jar:1.3:compile

Please let me know if I can give you more information.

Thank you for your help.

Best regards,
   Sascha Vujevic

**Comment by msymons** [comment]
The [FAQ](https://docs.dependencytrack.org/FAQ/) is a good place to start when trying to understand what is going on.  Taking things a bit further,  it is possible to investigate individual CVE.

Taking your first concern  (CVE-2021-29425) as an example, the component's "View Details"  in Dependency-Track gives you the PURL...

`pkg:maven/commons-io/commons-io@2.5?type=jar`

This can then be used to [query OSS Index](https://ossindex.sonatype.org/), which (as of 11th August 2021) reports:

> This version of commons-io has no known vulnerabilities!

This explains why the vulnerability is not yet showing up in Dependency-Track.  However, the vulnerability DOES exist...  so the next step is to log an issue in [OSSIndex vulns](https://github.com/OSSIndex/vulns).  Remembering, of course, to search the repo to double-check that something has not already been logged.

In this case it HAS been logged:  OSSIndex/vulns/issues/181

Observe my comment from last month!

Rinse and repeat for other CVE of concern.

**Comment by savujevi** [comment]
Thank you for your hint. I will check further information on the purl links.

I was only wondering why dependencycheck from OWASP did find the vulnerabilities. Is this the point of checking against different databases ?

Thanks.

**Reply by savujevi** [reply]
How do I know which Database is used ? If PURL writes pkg:... it is the OSS Index ?
Thank you.

**Reply by stevespringett** [reply]
OSS Index supports PURL. Snyk, which will be added in the near future supports PURL, as does other sources of vulnerability intelligence which will be supported in the future. PURL is also indirectly supported by NPM Advisories. There is logic to ecosystem-specific analysis. So if a PURL contains `npm` for a type, DT will automatically use NPM Audit APIs, and dynamically construct the payload that NPM expects, as it doesn't natively support PURL.

**Reply by stevespringett** [reply]
Also, for every vulnerability identified for a component, there is an attribution. That tells you exactly what datasource was used in the identification of the vulnerability
